File: articles/conservation-biology-protection-of-life/python/stochastic_pva.py
"""
Stochastic population viability analysis for conservation biology.

This script simulates population trajectories under environmental variability
and occasional catastrophe events. It estimates extinction and quasi-extinction
risk for a compact educational example.

Run:
    python python/stochastic_pva.py
"""

from __future__ import annotations

import numpy as np
import pandas as pd


def simulate_pva(
    initial_population: int = 120,
    years: int = 50,
    simulations: int = 1000,
    mean_growth_rate: float = 0.04,
    sd_growth_rate: float = 0.08,
    mean_carrying_capacity: float = 250.0,
    sd_carrying_capacity: float = 20.0,
    catastrophe_probability: float = 0.05,
    catastrophe_multiplier: float = 0.65,
    quasi_extinction_threshold: int = 20,
    random_seed: int = 42,
) -> dict[str, float | np.ndarray]:
    """Simulate stochastic population trajectories."""

    rng = np.random.default_rng(random_seed)
    trajectories = np.full((years + 1, simulations), np.nan)
    trajectories[0, :] = initial_population

    for simulation in range(simulations):
        population_size = float(initial_population)

        for year in range(1, years + 1):
            growth_rate = rng.normal(mean_growth_rate, sd_growth_rate)
            carrying_capacity = max(
                10.0,
                rng.normal(mean_carrying_capacity, sd_carrying_capacity),
            )

            # Stochastic logistic growth.
            population_size = population_size + growth_rate * population_size * (
                1.0 - population_size / carrying_capacity
            )

            # Catastrophe event such as drought, wildfire, disease, or storm impact.
            if rng.random() < catastrophe_probability:
                population_size *= catastrophe_multiplier

            population_size = max(0.0, round(population_size))
            trajectories[year, simulation] = population_size

            if population_size == 0:
                trajectories[year:, simulation] = 0
                break

    final_sizes = trajectories[-1, :]
    minimum_sizes = np.nanmin(trajectories, axis=0)

    return {
        "trajectories": trajectories,
        "extinction_risk": float(np.mean(final_sizes == 0)),
        "quasi_extinction_risk": float(
            np.mean(minimum_sizes <= quasi_extinction_threshold)
        ),
        "median_final_size": float(np.nanmedian(final_sizes)),
        "mean_final_size": float(np.nanmean(final_sizes)),
    }


def main() -> None:
    """Run the example simulation and print summary statistics."""

    result = simulate_pva()

    summary = pd.DataFrame(
        [
            {
                "extinction_risk": result["extinction_risk"],
                "quasi_extinction_risk": result["quasi_extinction_risk"],
                "median_final_size": result["median_final_size"],
                "mean_final_size": result["mean_final_size"],
            }
        ]
    )

    print(summary.round(3).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/conservation-biology-protection-of-life/r/stochastic_pva.R
# Stochastic population viability analysis in R.
#
# This script simulates many population trajectories under environmental
# variation and estimates extinction and quasi-extinction risk.

set.seed(42)

simulate_pva <- function(
  n0 = 120,
  years = 50,
  n_sims = 1000,
  r_mean = 0.04,
  r_sd = 0.08,
  k_mean = 250,
  k_sd = 20,
  catastrophe_prob = 0.05,
  catastrophe_mult = 0.65,
  quasi_extinction = 20
) {
  trajectories <- matrix(NA_real_, nrow = years + 1, ncol = n_sims)
  trajectories[1, ] <- n0

  for (sim in seq_len(n_sims)) {
    population_size <- n0

    for (year in seq_len(years)) {
      r_t <- rnorm(1, mean = r_mean, sd = r_sd)
      k_t <- max(10, rnorm(1, mean = k_mean, sd = k_sd))

      population_size <- population_size +
        r_t * population_size * (1 - population_size / k_t)

      if (runif(1) < catastrophe_prob) {
        population_size <- population_size * catastrophe_mult
      }

      population_size <- max(0, round(population_size))
      trajectories[year + 1, sim] <- population_size

      if (population_size == 0) {
        trajectories[(year + 1):(years + 1), sim] <- 0
        break
      }
    }
  }

  minimum_sizes <- apply(trajectories, 2, min, na.rm = TRUE)
  final_sizes <- trajectories[years + 1, ]

  list(
    trajectories = trajectories,
    extinction_risk = mean(final_sizes == 0, na.rm = TRUE),
    quasi_extinction_risk = mean(minimum_sizes <= quasi_extinction, na.rm = TRUE),
    median_final_size = median(final_sizes, na.rm = TRUE),
    mean_final_size = mean(final_sizes, na.rm = TRUE)
  )
}

results <- simulate_pva()

summary <- data.frame(
  extinction_risk = results$extinction_risk,
  quasi_extinction_risk = results$quasi_extinction_risk,
  median_final_size = results$median_final_size,
  mean_final_size = results$mean_final_size
)

print(round(summary, 3))
